app/services/google/parser_service.rb

```diff
@@ -0,0 +1,66 @@
+# frozen_string_literal: true
+
+module Google
+  class ParserService
+    NON_ADS_RESULT_SELECTOR = 'a[data-ved]:not([role]):not([jsaction]):not(.adwords):not(.footer-links)'
+    AD_CONTAINER_ID = 'tads'
+    ADWORDS_CLASS = 'adwords'
+
+    def initialize(html_response:)
+      raise ArgumentError, 'response.body cannot be blank' if html_response.body.blank?
+
+      @html = html_response
+
+      @document = Nokogiri::HTML.parse(html_response)
+
+      # Add a class to all AdWords link for easier manipulation
+      @document.css('div[data-text-ad] a[data-ved]').add_class(ADWORDS_CLASS)
+
+      # Mark footer links to identify them
+      @document.css('#footcnt a').add_class('footer-links')
+    end
+
+    # Write parsed data directly in the keyword object argument
+    # and return back the keyword object
+    def parse_into!(keyword)
+      keyword.ads_top_count = ads_top_count
+      keyword.ads_page_count = ads_page_count
+      keyword.ads_top_url = ads_top_url
+      keyword.ads_page_url = ads_page_url
```

Comment:
@malparty Service classes usually have just a 2 public methods `call` and `initialize`. Also, from this class it would be better to return the attributes in hash, instead of modify the keyword object.

Reply:
Nice, thanks to this feedback I discovered we can assign a hash values to ActiveRecord instance doing that:
```ruby
keyword.attributes = hash_of_attributes
``` 
I'll update the code accordingly ;-)

Reply:
Attribute methods set as private, `parse_into!` renamed in `call`, specs updated accordingly :) 
It should be good now ;-)